File: leet-code/maxSubArray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
  let pre = 0
  let ans = nums[0]
  for (let i = 0; i < nums.length; i++) {
    pre = Math.max(pre + nums[i], nums[i])
    ans = Math.max(pre, ans)
  }
  return ans
}

File: leet-code/maxSubArray.go
package main

func maxSubArray(nums []int) int {
	pre := 0
	res := nums[0]

	var max func(a int, b int) int
	max = func(a int, b int) int {
		if a > b {
			return a
		}
		return b
	}

	for i := range nums {
		pre = max(pre+nums[i], nums[i])
		res = max(pre, res)
	}

	return res
}
